src/main/java/org/myrobotlab/codec/CodecUtils.java

```diff
@@ -241,11 +246,11 @@ public static String capitalize(final String line) {
      * @param clazz The target class.
      * @param <T>   The type of the target class.
      * @return An object of the specified class (or a subclass of) with the state
-     * given by the json.
+     * given by the json. Null is an allowed return object.
      * @throws JsonDeserializationException if an error during deserialization occurs.
      * @see #USING_GSON
      */
-    public static <T extends Object> T fromJson(String json, Class<T> clazz) {
+    public static <T> @Nullable T fromJson(@Nonnull String json, @Nonnull Class<T> clazz) {
         try {

```

Comment:
`gson.fromString(null, Object.class)` returns null while `mapper.readValue(null, Object.class)` throws an exception, and the same goes for any target class. In my opinion, the difference between a JSON null value and an invalid input is that if the JSON contains null, the string should be `"null"` and not actually null. Passing in a null would indicate some failure upstream and shouldn't be taken as intentional

Reply:
null is valid json and null String is a valid reference to a non-existent string.  "null" is not null its just a string, and if we ever see "null" we probably have a problem to fix..
IMO null is valid input and a null input should be a null output - and should not throw.
A simple string is valid JSON too, and should not throw as input.
This is an interesting post describing the validity of a string only - https://stackoverflow.com/questions/7487869/is-this-simple-string-considered-valid-json.
I think null input should not throw, we should simply return null - and let downstream code handle the consequences

Reply:
> "null" is not null its just a string

I think this is where some confusion lies, it's correct that in `{"someval": "null"}`, `"null"` is just a string. But when it reaches `fromJson()`, that json is encoded as `"{\"someval\": \"null\"}"` in Java parlance. Likewise, the json value `null` is encoded in Java as `"null"`, so that `CodecUtils.toJson(null)` returns `"null"`, and not a null value (String containing `null` versus a `null` value entirely). If the JSON is intended to be null, we would pass `CodecUtils.fromJson("null", Object.class)`. This stuff definitely does get confusing, but if we look at Jackson we'll see it throws, and it seems to adhere pretty well to the JSON spec.

But, I can indeed allow that if it's what is desired, would just need to add a return in the catch

Reply:
The context of this is a message's method parameter data right ?
I'm thinking Java method signatures which that take Booleans, Integers, Strings, Doubles, and Objects certainly allow null.
If its the Message is json encoded the non-string objects are should not be in quotes.
Heh, they are when its double encoded.  But that's another topic that can make the null topic more complex :)

My concern is if the decoder throws more, are some methods might no longer accessible with null values :shrug: 

```
{
  "method": "someMethod", 
  "data":[null, 123]
}

public void someMethod(Boolean b, int) {

```
Honestly, I'm not positive of the behavior now ... but whatever we decide to do, we should figure out how it currently behaves - make a decision on it, and lock that decision with a unit test.

Reply:
> My concern is if the decoder throws more, are some methods might no longer accessible with null values shrug
The following should work without throwing...

That example would indeed work, because this:
```json
{
  "method": "someMethod", 
  "data":[null, 123]
}
```

Is a JSON-encoded String, including the null value embedded inside of it. There's already a test to ensure that works in CodecUtilsTest, the one called `testJsonMessageDeserNull`. Copied below for context:

```java
@Test
public void testJsonMessageDeserNull() {
    Message deserMessage = CodecUtils.jsonToMessage("null");
    assertNull(deserMessage);

}

```


When we pass JSON to the deserializer it is always in the form of a non-null String object, even if it represents a null value. Take the body of a POST request, for example: if you want the data to deserialize to null, you'd put `null` in the body, and that null is read as a String and passed to the deserializer. But if you put nothing, the body itself is null and will not deserialize at all because there's nothing there to deserialize.

Take this test here:
```java
Message deserMessage = CodecUtils.jsonToMessage(null);
```

Note how this is the same test mentioned above, but the null is not quoted, making it a null object instead of a String representing null. My intention is to make this throw because this is different from the test mentioned above and is only possible when there is a fault somewhere in the caller. This does not represent a null Message object like the previous test, but rather a complete lack of JSON at all, and interpreting it as a null Message can be dangerous.

Another example: passing `10` to `fromJson()` will not compile because `10` is not a String but rather an integer. But if we do this: `fromJson("10", Integer.class)`, we get the correct object. If we want to be consistent, then passing a null object should also fail because it is not a String, Java's type system just doesn't enforce that.

How it behaves right now differs depending on whether jackson or gson is used, that was unintentional, I thought the `@Nonnull` annotation would insert assertions to guarantee it's not null, but apparently, it does not do that. When GSON is selected, passing in a null object will return null, and when jackson is selected doing the same will throw an exception. Note that jackson is still able to deserialize messages that contain null data parameters, and the first test mentioned above succeeds with both. I will insert manual null checks / null returns once we come to a decision to keep the behavior between the two in parity.

Reply:
Cool, having a null parameter in as a method parameter is a common thing, and sounds like jacksons and the updates can handle that.  null as input as a json string to decode explodes.  As you said, it would a problem with the client. :+1: